pkg/services/dns/dns.go

```diff
@@ -85,116 +129,30 @@ func (h *dnsHandler) addAnswers(m *dns.Msg) {
 			}
 		}
 
-		resolver := net.Resolver{
-			PreferGo: false,
+		// need to create new message struct, as reusing original message struct leading
+		// to request errors
```

Comment:
I was quite confused by this for a long while, imo this is a good opportunity to rework `handle` and `addAnswers` .
The reason the copy is needed is because `r` is a "reply" message, sending it as a "query" message with `Exchange` won't work as expected. The copy you do will clear the `m.Response` boolean which is set to true.

Looking at `handle`, we have:
```
func (h *dnsHandler) handle(w dns.ResponseWriter, r *dns.Msg, responseMessageSize int) {
	m := new(dns.Msg)
	m.SetReply(r)
	m.RecursionAvailable = true
	h.addAnswers(m)
```
It gets a "query" message in the `r` parameter, it creates a "reply" message as `m`, and passes it to `addAnswers` so that `m.Answers` gets filled, and then the `m` "reply" is sent to back.

However, `addAnswers` does not limit itself to filling the `Answers` of the "reply" message. It used to reuse the `Query` field from that message to generate some go DNS lookups, and now with your changes, we need a full `dns.Msg` instance to forward to the system DNS server as a "query". So we have a "reply" message which we try to turn into a "query" message, which you did through this struct copying.

In my opinion, it would be cleaner to forward the initial request in `Exchange`, ie the `r` argument to the `handle` method, and to keep `m` strictly as a "reply" message.
I'd also argue that a lot of the confusion comes from the one-letter variables, `r`, `m`, ... :-/

I've done some work towards this in https://github.com/cfergeau/gvisor-tap-vsock/tree/replace-dns 
This includes a new test which is useful to test this code.
If you agree with the approach, feel free to take this code over and rework it in the way you like, the code in my branch is not final, just a way for me to check this can work :)

Reply:
I have add your code in this PR